hetmech/dwpc.py

```diff
@@ -204,15 +207,176 @@ def dwpc_long_repeat(graph, metapath, damping=0.5):
     raise NotImplementedError("See PR #59")
 
 
+def categorize(metapath):
+    """
+    Returns the classification of a given metapath as one of
+    a set of metapath types which we approach differently.
+    Parameters
+    ----------
+    metapath : hetio.hetnet.MetaPath
+
+    Returns
+    -------
+    classification : string
+        One of ['no_repeats', 'disjoint', 'short_repeat',
+                'long_repeat', 'BAAB', 'BABA', 'other']
+    Examples
+    --------
+    GbCtDlA -> 'no_repeats'
+    GiGiG   -> 'short_repeat'
+    GiGiGcG -> 'long_repeat'
+    GiGcGiG -> 'long_repeat'
+    GiGbCrC -> 'disjoint'
+    GbCbGbC -> 'BABA'
+    GbCrCbG -> 'BAAB'
+    DaGiGbCrC -> 'disjoint'
+    GiGaDpCrC -> 'disjoint'
+    GiGbCrCpDrD -> 'disjoint'
+    GbCpDaGbCpD -> NotImplementedError
+    GbCrCrCrCrCbG -> NotImplementedError
+    """
+    metanodes = list(metapath.get_nodes())
+    repeated_nodes = {v for i, v in enumerate(metanodes) if
+                      v in metanodes[i + 1:]}
+    freq = collections.Counter(metanodes)
+
+    if not repeated_nodes:
+        return 'no_repeats'
+
+    repeats_only = [node for node in metanodes if node in repeated_nodes]
+
+    # Group neighbors if they are the same
+    grouped = [list(v) for k, v in itertools.groupby(repeats_only)]
+
+    # Handle multiple disjoint repeats, any number, ie. AA,BB,CC,DD,...
+    if len(grouped) == len(repeated_nodes):
+        # Identify if there is only one metanode
+        if len(set(repeated_nodes)) == 1:
+            if max(freq.values()) < 4:
+                return 'short_repeat'
+            else:
+                return 'long_repeat'
+
+        return 'disjoint'
+
+    # Categorize the reformatted metapath
+    if len(repeats_only) in (3, 4):
+        if repeats_only[0] == repeats_only[-1]:
+            return 'BAAB'
+        else:
+            return 'BABA'
+    elif len(repeats_only) == 5 and max([len(i) for i in grouped]) == 3:
+        if repeats_only[0] == repeats_only[-1]:
+            return 'BAAB'
+    else:
+        # Multi-repeats that aren't disjoint, eg. ABCBAC
+        if len(repeated_nodes) > 2:
+            logging.info(
+                f"{metapath}: Only two overlapping repeats currently supported"
+            )
+
+        if len(metanodes) > 5:
+            logging.info(
+                f"{metapath}: Complex metapaths of length > 4 are not yet "
+                f"supported")
+        return 'other'
+
```

Comment:
But come to think of it, that may actually make the code in this region both more readable and easier to change in the future if either of the logged conditions is changed by a new DWPC function.